pkg/catalog/manager/catalog_sync.go

```diff
@@ -64,51 +64,68 @@ func (m *Manager) Sync(key string, obj *v3.Catalog) (runtime.Object, error) {
-		systemCatalogImageCache, err = m.ConfigMapLister.Get(namespace.System, systemCatalogImageCacheName)
-
-		// if the cache does not exist generate it
-		if err != nil && errors.IsNotFound(err) {
-			logrus.Debug("system catalog image cache configmap not found")
-
-			systemCatalogImageCache = &v1.ConfigMap{}
-			systemCatalogImageCache.Name = systemCatalogImageCacheName
-			systemCatalogImageCache.Namespace = namespace.System
-			err = image.CreateCatalogImageListConfigMap(systemCatalogImageCache, catalog)
-			if err != nil {
-				return nil, err
-			}
+		// ensure the system catalog image cache exists
+		forceUpdate := !isUpToDate(commit, catalog)
+		return nil, CreateOrUpdateSystemCatalogImageCache(catalog, m.ConfigMap, m.ConfigMapLister, false, forceUpdate)
+	}
 
-			systemCatalogImageCache, err = m.ConfigMap.Create(systemCatalogImageCache)
-			if err != nil && !errors.IsAlreadyExists(err) {
-				return nil, err
-			}
-			systemCatalogImageCacheCreated = true
-			logrus.Debug("system catalog image cache configmap created")
+	return nil, nil
+}
+
+func CreateOrUpdateSystemCatalogImageCache(systemCatalog *v3.Catalog, configMapInterface v1.ConfigMapInterface, configMapLister v1.ConfigMapLister, bundledMode bool, forceUpdate bool) (err error) {
+	var catalogChartPath string
+	catalogChartPath, err = utils.GetCatalogChartPath(systemCatalog, bundledMode)
+	if err != nil {
+		return err
+	}
+
+	var systemCatalogImageCache *v1.ConfigMap
+	var systemCatalogImageCacheCreated bool
+
+	systemCatalogImageCacheName := utils.GetCatalogImageCacheName(systemCatalog.Name)
+	systemCatalogImageCache, err = configMapLister.Get(namespace.System, systemCatalogImageCacheName)
 
-		} else {
-			return nil, err
+	// if the cache does not exist generate it
+	if err != nil && errors.IsNotFound(err) {
+		logrus.Debug("system catalog image cache configmap not found")
+
+		systemCatalogImageCache = &v1.ConfigMap{}
+		systemCatalogImageCache.Name = systemCatalogImageCacheName
+		systemCatalogImageCache.Namespace = namespace.System
+
+		err = image.AddImagesToImageListConfigMap(systemCatalogImageCache, catalogChartPath)
+		if err != nil {
+			return
 		}
 
-		// if the cache exists and is out of date update it
-		if !isUpToDate(commit, catalog) || systemCatalogImageCacheCreated {
-			err = image.CreateCatalogImageListConfigMap(systemCatalogImageCache, catalog)
-			if err != nil {
-				return nil, err
-			}
+		systemCatalogImageCache, err = configMapInterface.Create(systemCatalogImageCache)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			return
+		}
+		systemCatalogImageCacheCreated = true
+		logrus.Debug("system catalog image cache configmap created")
 
-			// update
-			_, err = m.ConfigMap.Update(systemCatalogImageCache)
-			if err != nil {
-				return nil, err
-			}
-			logrus.Debug("system catalog image cache updated")
+	} else if err != nil {
+		return
+	}
+
+	// if the cache exists and is out of date update it
+	if forceUpdate || systemCatalogImageCacheCreated {
```

Comment:
what's the use case for updating if systemCatalogImageCacheCreated=true, only forceUpdate=true should suffice?

Reply:
This was copied from the original function's logic in https://github.com/rancher/rancher/blob/fa481ddb14e3a3413ca5d75e709474bfb02a7eab/pkg/catalog/manager/catalog_sync.go#L98

I believe the edge case it catches is if the ConfigMap is created by another Rancher instance during a `Sync` on the catalog manager.

If a ConfigMap were to be created after this line:
https://github.com/rancher/rancher/blob/fa481ddb14e3a3413ca5d75e709474bfb02a7eab/pkg/catalog/manager/catalog_sync.go#L72

Then our Create call in the following line can return an `IsAlreadyExist`:
https://github.com/rancher/rancher/blob/fa481ddb14e3a3413ca5d75e709474bfb02a7eab/pkg/catalog/manager/catalog_sync.go#L86-L87

In this case, if we are doing a forceUpdate, then we should still upgrade the ConfigMap with the latest.

Reply:
Yeah, this might help resolve this, I think the create through any sync would have already fetched the latest images but makes sense to leave it as is from old logic than remove it.